src/System.Text.Json/src/System/Text/Json/JsonDocument.cs

```diff
@@ -0,0 +1,436 @@
+            ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
+
+            if (Utf8Parser.TryParse(segment, out long tmp, out int consumed) &&
+                consumed == segment.Length)
+            {
+                value = tmp;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal bool TryGetValue(int index, out ulong value)
+        {
+            if (_utf8Json.IsEmpty)
+                throw new ObjectDisposedException(nameof(JsonDocument));
+
+            _parsedData.Get(index, out DbRow row);
+
+            if (row.TokenType != JsonTokenType.Number)
+            {
+                throw new InvalidOperationException();
+            }
+
+            ReadOnlySpan<byte> data = _utf8Json.Span;
+            ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
+
+            if (Utf8Parser.TryParse(segment, out ulong tmp, out int consumed) &&
+                consumed == segment.Length)
+            {
+                value = tmp;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal bool TryGetValue(int index, out double value)
+        {
+            if (_utf8Json.IsEmpty)
+                throw new ObjectDisposedException(nameof(JsonDocument));
+
+            _parsedData.Get(index, out DbRow row);
+
+            if (row.TokenType != JsonTokenType.Number)
+            {
+                throw new InvalidOperationException();
+            }
+
+            ReadOnlySpan<byte> data = _utf8Json.Span;
+            ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
+
+            if (Utf8JsonReader.TryGetDoubleValue(segment, out double tmp))
+            {
+                value = tmp;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal string GetRawValueAsString(int index)
+        {
+            ReadOnlyMemory<byte> segment = GetRawValue(index);
+            return Encoding.UTF8.GetString(segment.Span);
+        }
+
+        private static void Parse(
+            ReadOnlySpan<byte> utf8JsonSpan,
+            Utf8JsonReader reader,
+            ref CustomDb database,
+            ref CustomStack stack)
+        {
+            bool inObject = false;
+            bool inArray = false;
+            int arrayItemsCount = 0;
+            int numberOfRowsForMembers = 0;
+            int numberOfRowsForValues = 0;
+            int parentLocation = -1;
+
+            ref byte jsonStart = ref MemoryMarshal.GetReference(utf8JsonSpan);
+
+            while (reader.Read())
+            {
+                JsonTokenType tokenType = reader.TokenType;
+
+                int tokenStart = Unsafe.ByteOffset(
+                    ref jsonStart,
+                    ref MemoryMarshal.GetReference(reader.ValueSpan)).ToInt32();
```

Comment:
Utf8JsonReader doesn't provide this information? If not, is that a flaw in the Utf8JsonReader API. that JsonDocument needs to resort to such tactics to get data Utf8JsonReader doesn't provide?